### PR TITLE
Use Docker Buildx and its new functionality for build performance

### PR DESCRIPTION
**Pull request by inductor**
Hi mattn,

I always admire your dajare. Kindly have a look what I've added for better `docker build` experience both on command line and GitHub Actions.

### CODE REVIEW

File: .github/workflows/docker.yaml [hunk 1]
```diff
@@ -2,33 +2,44 @@ name: Publish Docker image
...
       with:
         images: mattn/k8s-go-app
 
     - name: Build and push Docker image
       uses: docker/build-push-action@ad44023a93711e3deb337508980b4b5e9bcdc5dc
       with:
         context: .
-        push: true
+        builder: ${{ steps.buildx.outputs.name }}
+        push: ${{ github.ref == 'refs/heads/main' }}
```
Comment (inductor) [inline .github/workflows/docker.yaml:41]: This means if the push is on main it pushes to the container registry. If it's on some other branch it just tests build and doesn't push.

File: Dockerfile [hunk 2]
```diff
@@ -1,7 +1,9 @@
 FROM golang:1.18-alpine AS build-dev
 WORKDIR /go/src/app
-COPY . ./
+COPY --link go.mod go.sum ./
```
Comment (inductor) [inline Dockerfile:5]: Ref. https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/syntax.md#linked-copies-copy---link-add---link

File: README.md [hunk 3]
```diff
@@ -9,17 +9,19 @@ Small skeleton app running on k8s
 *NOTE*
 Please replace `${repositry}` in below and `k8s/*.yml` to point your private/public repositry.
 
-```
-docker build -t ${repositry}/k8s-go-app:latest .
-docker tag k8s-go-app:latest ${repositry}/k8s-go-app:latest
-docker push ${repositry}/k8s-go-app:latest
+```sh
+export repositry=docker.io/ # Modify
```
Comment (inductor) [inline README.md]: Looks better if you know where to modify(this won't change the script's behavior by default anyway)

File: .github/workflows/docker.yaml [hunk 4]
```diff
@@ -2,33 +2,44 @@ name: Publish Docker image
...
         tags: ${{ steps.meta.outputs.tags }}
         labels: ${{ steps.meta.outputs.labels }}
+        cache-from: type=gha
+        cache-to: type=gha,mode=max # mode=maxを有効にすると、中間ステージまで含めてキャッシュできる
```
Comment (inductor) [inline .github/workflows/docker.yaml:45]: Isn't this quite useful :D
https://github.com/docker/build-push-action/blob/master/docs/advanced/cache.md#cache-backend-api

**Comment by mattn** [comment]
This is helpful to gain my knowledges. Thanks.

**Comment by mattn** [comment]
Thank you!

**Comment by mattn** [comment]
I pushed tag v0.0.3 but tagged images seems not be available. Something wrong?

**Pull request merged**